src/stan/mcmc/hmc/xhmc/adapt_dense_e_xhmc.hpp

```diff
@@ -0,0 +1,55 @@
+#ifndef STAN_MCMC_HMC_xhmc_ADAPT_DENSE_E_XHMC_HPP
+#define STAN_MCMC_HMC_xhmc_ADAPT_DENSE_E_XHMC_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <stan/mcmc/stepsize_covar_adapter.hpp>
+#include <stan/mcmc/hmc/xhmc/dense_e_xhmc.hpp>
+
+namespace stan {
+  namespace mcmc {
+    /**
+      * The No-U-Turn sampler (XHMC) with multinomial sampling
+      * with a Gaussian-Euclidean disintegration and adaptive
+      * dense metric and adaptive step size
+    */
+    template <class Model, class BaseRNG>
+    class adapt_dense_e_xhmc : public dense_e_xhmc<Model, BaseRNG>,
+                               public stepsize_covar_adapter {
+    public:
+        adapt_dense_e_xhmc(Model &model, BaseRNG& rng)
+          : dense_e_xhmc<Model, BaseRNG>(model, rng),
+          stepsize_covar_adapter(model.num_params_r()) {}
+
+      ~adapt_dense_e_xhmc() {}
+
+      sample transition(sample& init_sample,
+                        interface_callbacks::writer::base_writer& writer) {
+        sample s = dense_e_xhmc<Model, BaseRNG>::transition(init_sample,
```

Comment:
I like this arrangement of calling (one of) the superclass(es) first and then doing adaptation.